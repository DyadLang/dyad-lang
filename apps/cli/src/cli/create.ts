import debug from "debug";
import path from "path";
import util from "util";
import { exec } from "child_process";
import { performCompile } from "./compile.js";
import { createLibrary } from "@juliacomputing/dyad-kernel";
import {
  dyadWatchTargets,
  NodeAsyncFileSystem,
} from "@juliacomputing/dyad-kernel/node";

export interface Options {
  debug: string;
  dir: string;
}

const execute = util.promisify(exec);

const cmdLog = debug("cli:create");

/**
 * This function creates a new Dyad library.  This effectively creates a new
 * Julia package but also adds some Dyad specific boilerplate as well.
 *
 * @param library The name of the library to create
 * @param options Options for the creation process
 */
export async function performLibraryCreation(
  library: string,
  options: Options
) {
  // If the debug option was enabled, enabled the debugging output channel.
  // This can also be done via the `DEBUG` environment variable (without the
  // command line flag).
  if (options.debug !== "") {
    debug.enable(options.debug);
  }
  let authors = { name: "", email: "" };
  try {
    authors = await author();
  } catch {
    cmdLog("Unable to determine current git user, author info will be blank");
  }
  const outputdir = path.join(options.dir, library);

  const nfs = new NodeAsyncFileSystem(outputdir, dyadWatchTargets);

  cmdLog("Creating library %s in %s", library, outputdir);
  await createLibrary(library, nfs, ".", authors);
  cmdLog("  ...library created, now performing initial compliation...");
  // Note: we use `project: "auto"` here because the library is guaranteed to have
  // no external dependencies (given that we have just created it).
  await performCompile(outputdir, { debug: options.debug, watch: false });
  cmdLog("  ...initial compilation finished");
}

async function author() {
  const name = await execute("git config --global user.name");
  const email = await execute("git config --global user.email");
  return {
    name: name.stdout.trim(),
    email: email.stdout.trim(),
  };
}
