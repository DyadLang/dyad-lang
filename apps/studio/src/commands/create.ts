import * as vscode from "vscode";
import { commandName } from "../common.js";
import { createLibrary } from "@juliacomputing/dyad-kernel";
import { NodeAsyncFileSystem } from "@juliacomputing/dyad-kernel/node";
import path from "path";

/**
 * This function registers the compile command
 * @param context
 * @param shared
 */
export function registerCreateComponentLibraryCommand(
  context: vscode.ExtensionContext
) {
  // The command has been defined in the package.json file
  // Now provide the implementation of the command with registerCommand
  // The commandId parameter must match the command field in package.json
  let disposable = vscode.commands.registerCommand(
    commandName("create"),
    async (args) => {
      await vscode.window.withProgress(
        {
          location: vscode.ProgressLocation.Window,
          cancellable: false,
          title: "Creating Library, please wait...",
        },
        async (progress) => {
          await performCreation(progress);
        }
      );

      console.log("Create arguments: ", args);
    }
  );
  context.subscriptions.push(disposable);
}

const validateLibraryName = new RegExp("^[a-zA-Z]+$");

async function performCreation(
  progress: vscode.Progress<{ message?: string; increment?: number }>
) {
  progress.report({ increment: 0 });

  try {
    // Prompt user for name of library and validate name
    const name = await vscode.window.showInputBox({
      title: "Component library name",
      value: "",
      placeHolder: "should end in 'Components'",
      validateInput: (value: string): string | null => {
        if (value.length < 5) {
          return "Library name must have at least 5 letters in it";
        }
        if (!validateLibraryName.test(value)) {
          return "Library name must contain only letters (upper or lowercase)";
        }
        return null;
      },
    });

    if (name === undefined) {
      await vscode.window.showErrorMessage("No library name provided");
      return;
    }

    progress.report({ increment: 25 });

    // Prompt user for directory to create library in
    let defaultDir: vscode.Uri | undefined = undefined;

    const folders = vscode.workspace.workspaceFolders;
    if (folders && folders.length > 0) {
      defaultDir = folders[0].uri;
    }

    const parentUri = await vscode.window.showOpenDialog({
      title: `Directory in which to create ${name} library`,
      defaultUri: defaultDir,
      canSelectFiles: false,
      canSelectFolders: true,
      canSelectMany: false,
    });

    if (parentUri === undefined) {
      await vscode.window.showErrorMessage("No directory selected");
      return;
    }

    const dir = path.join(parentUri[0].fsPath, name);

    if (dir === undefined) {
      await vscode.window.showErrorMessage("No directory provided");
      return;
    }

    progress.report({ increment: 50 });

    // Create library
    const start = performance.now();
    const fs = new NodeAsyncFileSystem(dir, []);
    await createLibrary(name, fs, ".", { name: "", email: "" });
    const created = performance.now();
    console.log(
      `Call to createLibrary took ${(created - start) / 1000} seconds`
    );

    progress.report({ increment: 75 });

    const updated = performance.now();
    console.log(`Updating settings took ${(updated - created) / 1000} seconds`);

    progress.report({ increment: 75 });

    const dirUri = vscode.Uri.joinPath(parentUri[0], name);

    // Open library in new Window
    vscode.commands
      .executeCommand("vscode.openFolder", dirUri, {
        forceNewWindow: true,
      })
      .then(
        () => {},
        (e) => {
          console.error(e);
        }
      );

    // // THIS DOESN'T REALLY RUN (the compilation at least)...because we open
    // // everything in a separate window.  Need to figure out a fix.
    // progress.report({ increment: 75 });

    // console.log("Library created");

    // // Now compile the code
    // await shared.client.caseOf({
    //   Nothing: async () => {
    //     await vscode.window.showErrorMessage(
    //       `Could not access initialized client`
    //     );
    //   },
    //   Just: async (client) => {
    //     // This sends the compile request to LSP.  This is a custom method but
    //     // we send it to the server because the Workspace is already set up on
    //     // that side.
    //     try {
    //       if (client.isRunning()) {
    //         await client.restart();
    //       }
    //       await triggerCompile(client);
    //     } catch (e) {
    //       console.error(`Error while compiling: `, e);
    //     }
    //   },
    // });
    // console.log("Completed library creation");
    // await vscode.window.showInformationMessage(
    //   `New Dyad Component library "${name}" created and compiled`
    // );
  } catch (e: any) {
    await vscode.window.showErrorMessage(e.message);
  } finally {
    progress.report({ increment: 100 });
  }
}
