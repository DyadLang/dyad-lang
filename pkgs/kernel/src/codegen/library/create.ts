import debug from "debug";
import path from "path";
import { v4 } from "uuid";
import { FileSystemInterface } from "../../providers/index.js";
import { PackageEntry, defaultContents, defaultPackages } from "./defaults.js";
import { sourceFolder } from "@juliacomputing/dyad-common";
import { formatDyad } from "@juliacomputing/dyad-parser";

const cmdLog = debug("codegen:library");

export interface CreateOptions {
  deps?: Record<string, PackageEntry>;
  contents?: Record<string, string>;
  uuid?: string;
}

export interface AuthorInformation {
  name: string;
  email: string;
}
const ALLOWLIST_FILES = new Set([".gitignore", ".git", ".DS_Store"]);
/**
 * This function creates a new Dyad library.  This effectively creates a new
 * Julia package but also adds some Dyad specific boilerplate as well.
 *
 * @param library The name of the library to create
 * @param fs File system to write files to
 * @param outputdir Directory on file system where file will be written
 * @param author Information on the author of this library
 */
export async function createLibrary(
  library: string,
  fs: FileSystemInterface,
  outputdir: string,
  author: AuthorInformation,
  options?: CreateOptions
) {
  cmdLog("Creating library %s in %s", library, outputdir);

  const contents = options?.contents || defaultContents;
  const deps = options?.deps
    ? { ...defaultPackages, ...options.deps }
    : defaultPackages;

  const exists = await fs.exists(outputdir, { type: "directory" });
  if (exists) {
    // Allow empty git repositories
    const contents = (await fs.readdir(outputdir)).filter(
      (x) => !ALLOWLIST_FILES.has(x)
    );
    if (contents.length > 0) {
      throw new Error(
        `Cannot create ${outputdir} because it already exists and is not empty`
      );
    }
  } else {
    await fs.mkdir(outputdir, { recursive: true });
  }

  const generated = path.join(outputdir, "generated");
  const src = path.join(outputdir, "src");
  const test = path.join(outputdir, "test");
  const sourceDir = path.join(outputdir, sourceFolder);
  const assets = path.join(outputdir, "assets");
  const project = path.join(outputdir, "Project.toml");

  const entry = path.join(src, `${library}.jl`);

  const toml = await projectToml(library, deps, author, options ?? {});
  for (const sub of [generated, src, test, sourceDir, assets]) {
    await fs.mkdir(sub);
  }

  const entryFile = await entryFileSource(library);

  // Put a .gitkeep file in assets
  await fs.writeFile(path.join(assets, ".gitkeep"), "");
  await fs.writeFile(path.join(outputdir, ".gitignore"), gitIgnore());
  await fs.writeFile(path.join(outputdir, ".gitattributes"), gitAttributes());
  await fs.writeFile(path.join(outputdir, "README.md"), readme(library));
  await fs.writeFile(project, toml);
  await fs.writeFile(entry, entryFile);
  await fs.writeFile(path.join(test, "runtests.jl"), testSource(library));
  await initializeVSCodeSettings(fs, outputdir);

  for (const [filename, source] of Object.entries(contents)) {
    const formatted = formatDyad(source);
    if (formatted.hasValue()) {
      await fs.writeFile(path.join(sourceDir, filename), formatted.value);
    } else {
      throw new Error(
        `Unable to properly format ${filename} when creating library ${library}...this should not happen`
      );
    }
  }
}

function gitAttributes() {
  return `/generated/** linguist-generated
/docs/src/components/** linguist-generated
/docs/src/tests/** linguist-generated
/docs/src/types/** linguist-generated
/test/snapshots/** linguist-generated
Manifest*.toml linguist-generated`;
}

function gitIgnore() {
  return `/docs/build/
*.jl.*.cov
*.jl.cov
*.jl.mem
/Manifest*.toml
/docs/Manifest*.toml
`;
}

async function projectToml(
  library: string,
  deps: Record<string, PackageEntry>,
  author: AuthorInformation,
  options: CreateOptions
) {
  const id = options?.uuid ?? v4();
  const user = author;
  const dsection: string[] = [];
  const csection: string[] = [];
  const esection: string[] = [];
  const targets: Set<string> = new Set();
  // Loop over all dependencies
  for (const [pname, entry] of Object.entries(deps)) {
    const restricted = entry.appliesTo && entry.appliesTo.length > 0;

    // Should this go in the main dependencies section?  If so, add it to
    // `dsection`, otherwise add it to the Extras section `esection`.
    if (restricted) {
      esection.push(`${pname} = "${entry.uuid}"`);
      for (const app of entry.appliesTo ?? []) {
        targets.add(app);
      }
    } else {
      dsection.push(`${pname} = "${entry.uuid}"`);
    }
    // Is there a compat for this, then add it to compats (`csection`)
    if (entry.compat) {
      csection.push(`${pname} = "${entry.compat}"`);
    }
  }
  const tlist = [...targets];
  tlist.sort();
  const tsection = tlist.map((target) =>
    Object.entries(deps)
      .filter((e) => e[1].appliesTo && e[1].appliesTo.includes(target))
      .map((e) => e[0])
  );

  return `name = "${library}"
uuid = "${id}"
authors = ["${user.name} <${user.email}>"]
version = "0.1.0"

[deps]
${dsection.join("\n")}

[compat]
${csection.join("\n")}

[extras]
${esection.join("\n")}

[targets]
${tlist.map((t, i) => `${t} = ${JSON.stringify(tsection[i])}`).join("\n")}
`;
}

function entryFileSource(library: string) {
  return `module ${library}

include("../generated/types.jl")
include("../generated/definitions.jl")
include("../generated/experiments.jl")
include("../generated/precompilation.jl")
    
end # module ${library}`;
}

function testSource(library: string) {
  return `
using ${library}
using Test
    
include("../generated/tests.jl")`;
}

function readme(library: string) {
  return `# ${library}
  
## Getting Started
  
This library was created with the Dyad Studio VS Code extension.  Your Dyad
models should be placed in the \`dyad\` directory and the files should be
given the \`.dyad\` extension.  Several such files have already been placed
in there to get you started.  The Dyad compiler will compile the Dyad models
into Julia code and place it in the \`generated\` folder.  Do not edit the
files in that directory or remove/rename that directory.

A complete tutorial on using Dyad Studio can be found [here](#).  But you
can run the provided example models by doing the following:

1. Run \`Julia: Start REPL\` from the command palette.

2. Type \`]\`.  This will take you to the package manager prompt.

3. At the \`pkg>\` prompt, type \`instantiate\` (this downloads all the Julia libraries
   you will need, and the very first time you do it it might take a while).

4. From the same \`pkg>\` prompt, type \`test\`.  This will test to make sure the models
   are working as expected.  It may also take some time but you should eventually
   see a result that indicates 2 of 2 tests passed.

5. Use the \`Backspace\`/\`Delete\` key to return to the normal Julia REPL, it should
   look like this: \`julia>\`.

6. Type \`using ${library}\`.  This will load your model library.

7. Type \`World()\` to run a simulation of the \`Hello\` model.  The first time you run it,
   this might take a few seconds, but each successive time you run it, it should be very fast.

8. To see simulation results type \`using Plots\` (and answer \`y\` if asked if you want
   to add it as a dependency).

9. To plot results of the \`World\` simulation, simply type \`plot(World())\`.

10. You can plot variations on that simulation using keyword arguments.  For example,
    try \`plot(World(stop=20, k=4))\`.
`;
}

async function initializeVSCodeSettings(
  fs: FileSystemInterface,
  outputdir: string
) {
  // Write settings into new directory **before opening it** This means the
  // Julia extension will automatically load the correct environment!
  await fs.mkdir(path.join(outputdir, ".vscode"));
  await fs.writeFile(
    path.join(outputdir, ".vscode", "settings.json"),
    JSON.stringify(
      {
        "julia.environmentPath": "${workspaceFolder}",
        "julia.NumThreads": "auto",
        "files.autoSave": "off",
      },
      null,
      4
    )
  );
}
