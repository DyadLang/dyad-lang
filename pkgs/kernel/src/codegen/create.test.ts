import { createLibrary } from "./library/create.js";
import { TestingFS } from "../workspace/utils.test.js";
import { fileExtension, sourceFolder } from "@juliacomputing/dyad-common";

describe("Test project creation", () => {
  test("Basic project", async () => {
    const fs = new TestingFS();
    await createLibrary("Basic", fs, ".", {
      name: "Test User",
      email: "test.user@juliahub.com",
    });

    expect(fs.directories).toEqual([
      ".",
      "generated",
      "src",
      "test",
      sourceFolder,
      "assets",
      ".vscode",
    ]);
    expect([...fs.files.keys()]).toEqual([
      "assets/.gitkeep",
      ".gitignore",
      ".gitattributes",
      "README.md",
      "Project.toml",
      "src/Basic.jl",
      "test/runtests.jl",
      ".vscode/settings.json",
      `${sourceFolder}/hello.${fileExtension}`,
    ]);

    const proj = fs.files.get("Project.toml");
    expect(proj).toBeDefined();
    if (proj !== undefined) {
      const lines = proj.split("\n");
      const withoutUUID = [lines[0], ...lines.slice(2)].join("\n");
      expect(withoutUUID).toMatchSnapshot();
      expect(
        fs.files.get(`${sourceFolder}/hello.${fileExtension}`)
      ).toMatchSnapshot();
    }
  });

  test("Overload initial contents", async () => {
    const fs = new TestingFS();
    await createLibrary(
      "Basic",
      fs,
      ".",
      {
        name: "Test User",
        email: "test.user@juliahub.com",
      },
      {
        contents: {
          "foo.dyad": `type Foo = Real`,
          "bar.dyad": "type Bar = Real",
        },
      }
    );

    expect([...fs.files.keys()]).toEqual([
      "assets/.gitkeep",
      ".gitignore",
      ".gitattributes",
      "README.md",
      "Project.toml",
      "src/Basic.jl",
      "test/runtests.jl",
      ".vscode/settings.json",
      `${sourceFolder}/foo.${fileExtension}`,
      `${sourceFolder}/bar.${fileExtension}`,
    ]);

    expect(fs.files.get(`${sourceFolder}/foo.${fileExtension}`)).toEqual(
      "type Foo = Real"
    );
    expect(fs.files.get(`${sourceFolder}/bar.${fileExtension}`)).toEqual(
      "type Bar = Real"
    );
  });

  test("Overload deps", async () => {
    const fs = new TestingFS();
    await createLibrary(
      "BasicWithDev",
      fs,
      ".",
      {
        name: "Test User",
        email: "test.user@juliahub.com",
      },
      {
        deps: {
          DiffEqDevTools: {
            uuid: "f3b72e0c-5b89-59e1-b016-84e28bfd966d",
            compat: "=2.45.0",
          },
        },
      }
    );

    const proj = fs.files.get("Project.toml");
    expect(proj).toBeDefined();
    if (proj !== undefined) {
      const lines = proj.split("\n");
      const withoutUUID = [lines[0], ...lines.slice(2)].join("\n");
      expect(withoutUUID).toMatchSnapshot();
    }
  });
});
