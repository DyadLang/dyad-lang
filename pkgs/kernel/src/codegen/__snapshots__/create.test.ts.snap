// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test project creation Basic project 1`] = `
"name = "Basic"
authors = ["Test User <test.user@juliahub.com>"]
version = "0.1.0"

[deps]
DyadEcosystemDependencies = "7bc808db-8006-421e-b546-062440d520b7"
Markdown = "d6f4376e-aef5-505a-96c1-9c027394607a"
DyadInterface = "99806f68-afab-45ca-9d8c-ceff6bc61f54"
ModelingToolkit = "961ee093-0014-501f-94e3-6117800e7a78"
OrdinaryDiffEqDefault = "50262376-6c5a-4cf5-baba-aaf4f84d72d7"
PrecompileTools = "aea7be01-6a6a-4083-8856-8a6e6704d82a"
RuntimeGeneratedFunctions = "7e49a35a-f44a-4d26-94aa-eba1b4ca6b47"

[compat]
DyadEcosystemDependencies = "=0.9.4"
Markdown = "1"

[extras]
CSV = "336ed68f-0bac-5ca0-87d4-7b16caf5d00b"
DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"

[targets]
test = ["CSV","DataFrames","Plots","Test"]
"
`;

exports[`Test project creation Basic project 2`] = `
"# A simple lumped thermal model
component Hello
  # Ambient temperature
  parameter T_inf::Temperature = 300
  # Initial temperature
  parameter T0::Temperature = 320
  # Convective heat transfer coefficient
  parameter h::CoefficientOfHeatTransfer = 0.7
  # Surface area
  parameter A::Area = 1.0
  # Mass of thermal capacitance
  parameter m::Mass = 0.1
  # Specific Heat
  parameter c_p::SpecificHeatCapacity = 1.2
  variable T::Temperature
relations
  # Specify initial conditions
  initial T = T0
  # Newton's law of cooling/heating
  m*c_p*der(T) = h*A*(T_inf-T)
metadata {"Dyad": {"tests": {"case1": {"stop": 10, "expect": {"initial": {"T": 320}}}}}}
end

analysis World
  extends TransientAnalysis(stop=10)
  model = Hello(T_inf=T_inf, h=h)
  parameter T_inf::Temperature = 300
  parameter h::CoefficientOfHeatTransfer = 0.7
end"
`;

exports[`Test project creation Overload deps 1`] = `
"name = "BasicWithDev"
authors = ["Test User <test.user@juliahub.com>"]
version = "0.1.0"

[deps]
DyadEcosystemDependencies = "7bc808db-8006-421e-b546-062440d520b7"
Markdown = "d6f4376e-aef5-505a-96c1-9c027394607a"
DyadInterface = "99806f68-afab-45ca-9d8c-ceff6bc61f54"
ModelingToolkit = "961ee093-0014-501f-94e3-6117800e7a78"
OrdinaryDiffEqDefault = "50262376-6c5a-4cf5-baba-aaf4f84d72d7"
PrecompileTools = "aea7be01-6a6a-4083-8856-8a6e6704d82a"
RuntimeGeneratedFunctions = "7e49a35a-f44a-4d26-94aa-eba1b4ca6b47"
DiffEqDevTools = "f3b72e0c-5b89-59e1-b016-84e28bfd966d"

[compat]
DyadEcosystemDependencies = "=0.9.4"
Markdown = "1"
DiffEqDevTools = "=2.45.0"

[extras]
CSV = "336ed68f-0bac-5ca0-87d4-7b16caf5d00b"
DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
Plots = "91a5bcdd-55d7-5caf-9e0b-520d859cae80"
Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"

[targets]
test = ["CSV","DataFrames","Plots","Test"]
"
`;
